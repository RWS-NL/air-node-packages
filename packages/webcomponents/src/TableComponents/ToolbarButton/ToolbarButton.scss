@import "../../styles/vars";

.paper {
  align-items: center;
  display: flex;
  height: 2.5rem;
  padding: 2px;
}

.input {
  flex: 1;
  font-size: 1rem;
  margin-left: 8px;

  &Button {
    padding: 10px;
  }
}
