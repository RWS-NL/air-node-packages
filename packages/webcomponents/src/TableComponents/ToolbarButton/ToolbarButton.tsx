import React, { FC, useMemo, memo } from 'react';
import { Paper, IconButton } from '@material-ui/core';
import css from './ToolbarButton.scss';
import classnames from 'classnames';
import Tooltip, { TooltipProps } from '../../Tooltip/Tooltip';
import { objectHasProperty } from '@rws-air/utils';

export interface ToolbarButtonProps {
  /** The Material Icon to display in this paper button */
  icon: JSX.Element;
  /**
   * Whether this button should be disabled or not
   * @default false
   */
  disabled?: boolean;
  /**
   * The text content to put in the tooltip
   * @default 'Click Me'
   */
  tooltipText?: TooltipProps['title'];
  /**
   * The placement for the tooltip
   * @default top
   */
  tooltipPlacement?: TooltipProps['placement'];
  /**
   * Disables the inclusion of a tooltip
   * @default false
   */
  disableTooltip?: boolean;
  /**
   * The elevation to apply to the paper component
   * @default 2
   */
  paperElevation?: number;
  /** Data-qa tag to apply to the paper component */
  'paper-data-qa'?: string;
  /** Custom CSS classes to pass to the paper component */
  paperCustomclasses?: string | string[];
  /** Data-qa tag to apply to the IconButton component */
  'iconButton-data-qa'?: string;
  /** Custom CSS classes to pass to the IconButton component */
  iconButtonCustomclasses?: string | string[];
  /** Custom CSS classes to pass to the Tooltip component */
  tooltipCustomClasses?: string | string[];
  /** Data-qa tag to apply to the Tooltip component */
  'tooltip-data-qa'?: string;
  /** The action to trigger when clicking the icon button */
  onClick(...args: unknown[]): unknown;
}

/**
 * Constructs a ToolbarButton (elevated button with Material Paper) using pre-defined Rijkswaterstaat styling
 * @param props Props to pass to the ToolbarButton
 * @example
 * <ToolbarButton
 *   icon={<CloudDownload/>}
 *   disabled={false}
 *   onClick={console.log}
 *   tooltipText='SAMPLE'
 *   tooltipPlacement={'top'}
 *   disableTooltip={false}
 * />
 */
export const ToolbarButton: FC<ToolbarButtonProps> = props => {
  const elevation = useMemo(
    () => objectHasProperty(props, 'paperElevation') ? props.paperElevation : 2,
    [ props.paperElevation ]
  );

  const renderWithTooltip = useMemo(
    () => (
      <Tooltip
        data-qa={props['tooltip-data-qa']}
        customclasses={props.tooltipCustomClasses}
        placement={props.tooltipPlacement || 'top'}
        title={props.tooltipText || 'Click Me'}
      >
        <IconButton
          disabled={props.disabled}
          data-qa={props['iconButton-data-qa']}
          className={classnames(css.inputButton, props.iconButtonCustomclasses)}
          onClick={props.onClick}
          size='small'
          color='primary'
        >
          {props.icon}
        </IconButton>
      </Tooltip>
    ),
    [
      props.disabled, props.onClick, props.icon, props.iconButtonCustomclasses,
      props.tooltipCustomClasses, props.tooltipPlacement, props.tooltipText
    ]
  );

  const renderWithoutTooltip = useMemo(() => (
    <IconButton
      disabled={props.disabled}
      data-qa={props['iconButton-data-qa']}
      className={classnames(css.inputButton, props.iconButtonCustomclasses)}
      onClick={props.onClick}
      size='small'
      color='primary'
    >
      {props.icon}
    </IconButton>
  ), [ props.disabled, props.onClick, props.icon, props.iconButtonCustomclasses ]);

  const customRender = useMemo(() => {
    if (props.disabled) return renderWithoutTooltip;
    if (props.disableTooltip) return renderWithoutTooltip;

    return renderWithTooltip;
  }, [ props.disabled, props.disableTooltip ]);

  return (
    <Paper
      className={classnames(css.paper, props.paperCustomclasses)}
      elevation={elevation}
      data-qa={props['paper-data-qa']}
      square
    >
      {customRender}
    </Paper>
  );
};

export default memo(ToolbarButton);