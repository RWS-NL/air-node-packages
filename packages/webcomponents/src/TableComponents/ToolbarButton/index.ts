export { default as ToolbarButton, ToolbarButtonProps } from './ToolbarButton';