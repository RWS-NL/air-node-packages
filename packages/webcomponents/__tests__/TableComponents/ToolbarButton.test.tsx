import { IconButton } from '@material-ui/core';
import SearchIcon from '@material-ui/icons/Search';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';
import ToolbarButton, { ToolbarButtonProps } from '../../src/TableComponents/ToolbarButton/ToolbarButton';

const mockButtonClick = jest.fn();

describe('Component Tests', () => {
  let tableToolbarButton: ShallowWrapper<ToolbarButtonProps>;

  beforeAll(() => {
    tableToolbarButton = shallow(<ToolbarButton icon={<SearchIcon />} onClick={mockButtonClick} />);
  });

  test('button functionality called on click', () => {
    const button = tableToolbarButton.find(IconButton);
    button.simulate('click');

    expect(mockButtonClick).toHaveBeenCalledWith();
    expect(mockButtonClick).toHaveBeenCalledTimes(1);
  });
});

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const toolbarButton = shallow(<ToolbarButton icon={<SearchIcon />} onClick={mockButtonClick} />);
    expect(toolbarButton).toMatchSnapshot();
  });

  test('Optional Props', () => {
    const toolbarButton = shallow(
      <ToolbarButton icon={<SearchIcon />} onClick={mockButtonClick} tooltipText='tooltip' tooltipPlacement='bottom' paperElevation={5} />
    );
    expect(toolbarButton).toMatchSnapshot();
  });

  test('Disabled Tooltip', () => {
    const toolbarButton = shallow(<ToolbarButton icon={<SearchIcon />} onClick={mockButtonClick} disableTooltip />);
    expect(toolbarButton).toMatchSnapshot();
  });
});