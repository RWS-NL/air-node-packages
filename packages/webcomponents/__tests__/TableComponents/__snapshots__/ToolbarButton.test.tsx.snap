// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Disabled Tooltip 1`] = `
<WithStyles(ForwardRef(Paper))
  className="paper"
  elevation={2}
  square={true}
>
  <WithStyles(ForwardRef(IconButton))
    className="inputButton"
    color="primary"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    size="small"
  >
    <SearchIcon />
  </WithStyles(ForwardRef(IconButton))>
</WithStyles(ForwardRef(Paper))>
`;

exports[`Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className="paper"
  elevation={5}
  square={true}
>
  <Memo()
    placement="bottom"
    title="tooltip"
  >
    <WithStyles(ForwardRef(IconButton))
      className="inputButton"
      color="primary"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      size="small"
    >
      <SearchIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Memo()>
</WithStyles(ForwardRef(Paper))>
`;

exports[`Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Paper))
  className="paper"
  elevation={2}
  square={true}
>
  <Memo()
    placement="top"
    title="Click Me"
  >
    <WithStyles(ForwardRef(IconButton))
      className="inputButton"
      color="primary"
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      size="small"
    >
      <SearchIcon />
    </WithStyles(ForwardRef(IconButton))>
  </Memo()>
</WithStyles(ForwardRef(Paper))>
`;
